src/uct/sm/mm/mm_iface.c

```diff
@@ -301,17 +342,30 @@ static UCS_CLASS_INIT_FUNC(uct_mm_iface_t, uct_pd_h pd, uct_worker_h worker,
                                     "mm_recv_desc", &self->recv_desc_mp);
     if (status != UCS_OK) {
         ucs_error("Failed to create a receive descriptor memory pool for the MM transport");
-        goto err;
+        goto err_free_fifo;
     }
 
     /* set the first receive descriptor */
     self->last_recv_desc = ucs_mpool_get(self->recv_desc_mp);
+    VALGRIND_MAKE_MEM_DEFINED(self->last_recv_desc, sizeof(*(self->last_recv_desc)));
     if (self->last_recv_desc == NULL) {
         ucs_error("Failed to get the first receive descriptor");
         status = UCS_ERR_NO_RESOURCE;
         goto destroy_recv_mpool;
     }
-    VALGRIND_MAKE_MEM_DEFINED(self->last_recv_desc, sizeof(*(self->last_recv_desc)));
+
+    /* initiate the owner bit in all the FIFO elements and assign a receive descriptor
+     * per every FIFO element */
+    for (i = 0; i < mm_config->fifo_size; i++) {
+        fifo_elem_p = UCT_MM_IFACE_GET_FIFO_ELEM(self, self->recv_fifo_elements, i);
+        fifo_elem_p->flags = UCT_MM_FIFO_ELEM_FLAG_OWNER;
+
+        status = uct_mm_assign_desc_to_fifo_elem(self, fifo_elem_p);
+        if (status != UCS_OK) {
+            ucs_error("Failed to allocate a descriptor for MM");
+            goto destroy_descs;
```

Comment:
it goes to freeing them right after the first one fails which means that all the previous ones were allocated. and it frees up to the failed one. why would it segfault?